[MKL-DNN] Fix to int8 performance regression (#18698)

jczaja opened the pull request:
Changes here are restoring performance of int8 convolution caused by muli-threading mechanisms introduction (issue #18698). 

Internal testing proved that there is performance improvement for int8 convolutional models

jczaja commented:
@luotao1 Could you please advise on why PR_CI_inference did not start yet?

luotao1 commented:
It seems that this PR passes all the CI now. Next time, you can close and reopen your PR to trigger the CI.

jczaja commented:
@luotao1 Thanks for suggesstion

Sand3r- commented on paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:
```diff
@@ -392,12 +392,24 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
+      auto tid = std::this_thread::get_id();
+      std::stringstream ss;
+      ss << tid;
+      key_tid = "-t:" + ss.str();
```
Please consider simplifying
```
std::stringstream ss;
ss << tid;
key_tid = "-t:" + ss.str();
```
down to:
`key_tid = "-t:" + std::to_string(tid);`

jczaja replied:
Thanks for suggestion. This is very elegant proposal. I'm not planning to change it in that PR as this code will be removed once @lidanqing-intel finish her unification task of convolution. But I have similar section of code inside of MKLDNNHandler , so I can experiment with your proposal in next PR

jczaja commented:
@luotao1 Could you please review this PR?

luotao1 commented:
Got it. I will test it ASAP.

luotao1 commented:
I test it on 6271 cpu with Mobilenet-V1/V2 only.
```
./test_analyzer_int8_image_classification --infer_data=int8_full_val.bin --warmup --batch_size=1 --paddle_num_threads=1 --infer_model=int8v2/mobilenetv1/model/
```
### Mobilenet-v1
- before:
```
I0725 13:02:23.514694 341334 tester_helper.h:462] --- Performance summary ---
I0725 13:02:23.514722 341334 tester_helper.h:463] FP32: avg fps: 80.6569, avg latency: 12.3982 ms
I0725 13:02:23.514744 341334 tester_helper.h:466] INT8: avg fps: 197.8102, avg latency: 5.0554 ms
I0725 13:02:23.524044 341334 tester_helper.h:447] --- Accuracy summary ---
I0725 13:02:23.524067 341334 tester_helper.h:448] Accepted top1 accuracy drop threshold: 0.01. (condition: (FP32_top1_acc - INT8_top1_acc) <= threshold)
I0725 13:02:23.524085 341334 tester_helper.h:451] FP32: avg top1 accuracy: 0.7078
I0725 13:02:23.524091 341334 tester_helper.h:453] INT8: avg top1 accuracy: 0.7041
```
- after:
```
I0725 14:03:20.350790 92589 tester_helper.h:462] --- Performance summary ---
I0725 14:03:20.350821 92589 tester_helper.h:463] FP32: avg fps: 77.4833, avg latency: 12.9060 ms
I0725 14:03:20.350841 92589 tester_helper.h:466] INT8: avg fps: 219.8935, avg latency: 4.5477 ms
I0725 14:03:20.357447 92589 tester_helper.h:447] --- Accuracy summary ---
I0725 14:03:20.357475 92589 tester_helper.h:448] Accepted top1 accuracy drop threshold: 0.01. (condition: (FP32_top1_acc - INT8_top1_acc) <= threshold)
I0725 14:03:20.357496 92589 tester_helper.h:451] FP32: avg top1 accuracy: 0.7078
I0725 14:03:20.357501 92589 tester_helper.h:453] INT8: avg top1 accuracy: 0.7041
```
### Mobilenet-v2
- before
```
I0725 14:57:16.504341 247221 helper.h:322] ====== threads: 1, thread id: 0 ======
I0725 14:57:16.504350 247221 helper.h:324] ====== batch size: 1, iterations: 50000, repetitions: 1 ======
I0725 14:57:16.504357 247221 helper.h:326] ====== batch latency: 6.13307ms, number of samples: 50000, sample latency: 6.13307ms, fps: 163.05, data type: int8_t ======
I0725 14:57:16.509524 247221 tester_helper.h:462] --- Performance summary ---
I0725 14:57:16.509562 247221 tester_helper.h:463] FP32: avg fps: 93.0092, avg latency: 10.7516 ms
I0725 14:57:16.509582 247221 tester_helper.h:466] INT8: avg fps: 163.0504, avg latency: 6.1331 ms
I0725 14:57:16.518700 247221 tester_helper.h:447] --- Accuracy summary ---
I0725 14:57:16.518733 247221 tester_helper.h:448] Accepted top1 accuracy drop threshold: 0.01. (condition: (FP32_top1_acc - INT8_top1_acc) <= threshold)
```
- after
```
I0725 14:40:05.994760 368052 tester_helper.h:462] --- Performance summary ---
I0725 14:40:05.994797 368052 tester_helper.h:463] FP32: avg fps: 95.5066, avg latency: 10.4705 ms
I0725 14:40:05.994822 368052 tester_helper.h:466] INT8: avg fps: 191.5456, avg latency: 5.2207 ms
I0725 14:40:06.001505 368052 tester_helper.h:447] --- Accuracy summary ---
I0725 14:40:06.001518 368052 tester_helper.h:448] Accepted top1 accuracy drop threshold: 0.01. (condition: (FP32_top1_acc - INT8_top1_acc) <= threshold)
I0725 14:40:06.001543 368052 tester_helper.h:451] FP32: avg top1 accuracy: 0.7190
I0725 14:40:06.001561 368052 tester_helper.h:453] INT8: avg top1 accuracy: 0.7134
```

luotao1 commented:
It seems the `before` performance is better than https://github.com/PaddlePaddle/Paddle/issues/18698#issue-470205380. Is this normal @bingyanghuang @jczaja

bingyanghuang commented:
> It seems the `before` performance is better than [#18698 (comment)](https://github.com/PaddlePaddle/Paddle/issues/18698#issue-470205380). Is this normal @bingyanghuang @jczaja

So it seems that the performance regression problem has **not totally** been solved. @jczaja

jczaja commented:
@luotao1 I have made some measures using mobilnet v1 on SKX (8180), I will share them , perhaps this will help in understanding situation we are dealing with. 

1. changes before regression (one before commit 84bb45c ). 
batch latency: 6.12171ms, number of samples: 50000, sample latency: 6.12171ms, **fps: 163.353**, data type: int8_t
2. regression commit ( commit 84bb45c ):
 batch latency: 8.22298ms, number of samples: 50000, sample latency: 8.22298ms, **fps: 121.61**, data type: int8_t 
3. regression commit + this PR (ported changes to regression commit):
batch latency: 6.26656ms, number of samples: 50000, sample latency: 6.26656ms, **fps: 159.577**, data type: int8_t

4. Recent develop ( 25th of July):
 INT8: **avg fps: 150.7610**, avg latency: 6.6330 ms    

5. Recent develop + PR :
INT8: **avg fps: 159.3892**, avg latency: 6.2740 ms 

Comments:
a) From point 1 and 2 we can see that regression was from 163 fps -> 121 fps . Then in 3
we see that improvement is 121 fps -> 159 fps . Why not 163 ? Answer is that generation of string from Thread Id via stringstream caused 4 FPS penalty. To prove that I replaced stringstream with fixed value string (I have run analyzer in single threaded, so it was functionally fine). And performance was again 163 fps. To combat this problem I have updated this PR replacing stringstream usage with to_string(std::hash<std::thread::id>)() to get string out of TID.
I think there is small improvement 159 -> ~160 . 

b) You asked why recent before is much better that the one reported. If you look into result from point 2 (regression commit)  and recent develop (regression present, but recent codebase) we can see that there is an improvement 121 fps -> 150 fps. Why is that? I haven't traced it down, but possible reason is that we updated to mkl-dnn 0.20 , and mkl-dnn team reduced time of creation of primitives. So then performance gain is not that big with caching when overhead from creation is smaller. This is hypothesis (I looked at some of their commits, presentations so I know they worked on that) to be confirmed. 

c) From 4 and 5 we can see that there is improvement 150 fps -> 159 fps. Those results are taken on SKX , on next generation (CLX) that you are using difference should be a bit bigger.

Please continue reviewing those changes.

jczaja commented:
@luotao1 Could you please advise on how to pass: "PR_CI_Coverage" and "PR_CI_Document_Preview"

luotao1 commented:
You can ignore these two CI since they are not required now.

jczaja commented:
@luotao1 Please review

luotao1 commented:
- mobilenetV1: from 219.8935 (https://github.com/PaddlePaddle/Paddle/pull/18758#issuecomment-514914670) ->226.5409, [release1.5](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/tests/api/int8_mkldnn_quantization.md#2-accuracy-and-performance-benchmark) is 230.65.
```
I0730 12:18:00.623914 357906 tester_helper.h:463] FP32: avg fps: 78.4434, avg latency: 12.7480 ms
I0730 12:18:00.623934 357906 tester_helper.h:466] INT8: avg fps: 226.5409, avg latency: 4.4142 ms
```
- mobilenetv2: from 191.5456 (https://github.com/PaddlePaddle/Paddle/pull/18758#issuecomment-514914670) -> 201.2669, [release1.5](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/tests/api/int8_mkldnn_quantization.md#2-accuracy-and-performance-benchmark) is 206.92.
```
I0730 12:36:33.656062 269110 tester_helper.h:463] FP32: avg fps: 91.1440, avg latency: 10.9717 ms
I0730 12:36:33.656085 269110 tester_helper.h:466] INT8: avg fps: 201.2669, avg latency: 4.9685 ms
```
The test result is align with https://github.com/PaddlePaddle/Paddle/pull/18758#issuecomment-515482597.

luotao1 commented:
LGTM！And the last question 
> Why not 163 ? Answer is that generation of string from Thread Id via stringstream caused 4 FPS penalty. 

Will it be solved after refactoring conv Op?

jczaja commented:
@luotao1 Key generation using TID is not related to conv refactoring.  But I would like to try to speed it up , and this effort will be tracked with #18902

luotao1 commented:
Got it. And this PR could be merged at first!
